service.go

```diff
@@ -0,0 +1,541 @@
+		},
+		ThreadID: &pb.ProtoThreadID{ID: id},
+		LogID:    &pb.ProtoPeerID{ID: lid},
+		Offset:   &pb.ProtoCid{Cid: offset},
+		Limit:    int32(settings.Limit),
+	}
+
+	// Pull from each address
+	recs := newRecords()
+	wg := sync.WaitGroup{}
+	for _, addr := range s.threads.LogInfo(id, lid).Addrs {
+		wg.Add(1)
+		go func(addr ma.Multiaddr) {
+			p, err := addr.ValueForProtocol(ma.P_P2P)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			pid, err := peer.IDB58Decode(p)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Debugf("pulling records from %s...", p)
+
+			cctx, _ := context.WithTimeout(ctx, reqTimeout)
+			conn, err := s.dial(cctx, pid, grpc.WithInsecure())
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			client := pb.NewThreadsClient(conn)
+			reply, err := client.Pull(cctx, req)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			for _, r := range reply.Records {
+				rec, err := cbor.RecordFromProto(r, fk)
+				if err != nil {
+					log.Error(err)
+					return
+				}
+				recs.Store(rec.Cid(), rec)
+			}
+
+			log.Debugf("reply from %s: %s", p, reply.String())
+
+			wg.Done()
+		}(addr)
+	}
+
+	wg.Wait()
+	return recs.List(), nil
+}
+
+// dial attempts to open a GRPC connection over libp2p to a peer.
+func (s *service) dial(ctx context.Context, peerID peer.ID, dialOpts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	opts := append([]grpc.DialOption{s.getDialOption()}, dialOpts...)
+	return grpc.DialContext(ctx, peerID.Pretty(), opts...)
+}
+
+// getDialOption returns the WithDialer option to dial via libp2p.
+func (s *service) getDialOption() grpc.DialOption {
+	return grpc.WithContextDialer(func(ctx context.Context, peerIDStr string) (net.Conn, error) {
+		id, err := peer.IDB58Decode(peerIDStr)
+		if err != nil {
+			return nil, fmt.Errorf("grpc tried to dial non peer-id: %s", err)
+		}
+		c, err := gostream.Dial(ctx, s.threads.host, id, IPELProtocol)
+		return c, err
+	})
+}
+
+// subscribe to a thread for updates.
+func (s *service) subscribe(id thread.ID) {
+	sub, err := s.pubsub.Subscribe(id.String())
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	for {
+		msg, err := sub.Next(s.threads.ctx)
+		if err != nil {
+			log.Error(err)
+			break
+		}
+
+		req := new(pb.PushRequest)
+		err = proto.Unmarshal(msg.Data, req)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+
+		reply, err := s.Push(s.threads.ctx, req)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		log.Debugf("received multicast request (reply: %t)", reply.Ok)
+	}
+}
+
+// publish a request to a thread.
+func (s *service) publish(id thread.ID, req *pb.PushRequest) error {
+	data, err := req.Marshal()
+	if err != nil {
+		return err
+	}
+	return s.pubsub.Publish(id.String(), data)
+}
+
+// handleInvite attempts to process a log record as an invite event.
+func (s *service) handleInvite(ctx context.Context, id thread.ID, lid peer.ID, kid peer.ID, rec thread.Record) (*thread.LogInfo, error) {
```

Comment:
As the doc string says, we try to process the record as an invite... adding the logs it contains if valid.